dataflow/src/main/java/org/checkerframework/dataflow/util/IdentityMostlySingleton.java

```diff
@@ -21,21 +22,30 @@ public IdentityMostlySingleton(T value) {
 
     @Override
     @SuppressWarnings("fallthrough")
-    public boolean add(T e) {
+    public boolean add(@FindDistinct T e) {
         switch (state) {
             case EMPTY:
                 state = State.SINGLETON;
                 value = e;
                 return true;
             case SINGLETON:
+                if (value == e) {
+                    return false;
+                }
                 state = State.ANY;
+                // Use ArrayList, but make sure to use reference comparisons.
                 set = new ArrayList<>();
```

Comment:
I am partial to the version in #3550 because the variable named `set` is a set, and because it's shorter -- it doesn't need to re-implement logic for `add` and `contains` and therefore I find it easier to read and understand.  As a minor benefit, #3550 type-checks under the Nullness Checker without the need for assumptions about use of the `set` variable in other subclasses of `AbstractMostlySingleton` (namely, `@MonotonicNonNull` on that variable).